### PR TITLE
[RDY] UI prototypes

### PR DESCRIPTION
**Pull request by tarruda**
Prototyping a graphical and text UI for helping me fix issues with the abstract ui code. This branch depends on neovim/neovim#1657.

### CODE REVIEW

**Comment by bfredl** [comment]
Nice, input (including altGr) works very well. A slight discrepancy from teminal nvim though: clicking any word will visually select the entire word, unlike the mouse in the terminal which will just move the cursor in normal mode. (It looks like every press becomes a doubleclick)

**Comment by tarruda** [comment]
>  A slight discrepancy from teminal nvim though: clicking any word will visually select the entire word, unlike the mouse in the terminal which will just move the cursor in normal mode. (It looks like every press becomes a doubleclick)

Must be some timing issue, I just tried double clicking with a relatively high interval and it selects the word, like in terminal vim. FWIW the behavior is tested [here](https://github.com/neovim/neovim/blob/master/test/functional/ui/mouse_spec.lua#L94-L103)

**Comment by tarruda** [comment]
@bfredl I'm done with the gtk UI, do you mind giving it a try and check for pending issues? The mouse click issue should be fixed now.

Now I'm gonna prototype a libtickit UI using cffi, and then translate it to C in neovim core.

**Comment by bfredl** [comment]
The issue of missing formatting I had before is now gone, but I still have the mouse-click issue. (This also doesn't support external input methods yet, but I have a [branch locally](https://github.com/bfredl/python-client/commit/2425fe9f32e3f7d3702370cdfc5803fe293a90b8) that mostly implements this. I will push PR when this is done)

**Comment by tarruda** [comment]
Done with this PR, I will now perform some final adjustments in neovim/neovim#1657 and then merge everything.

File: neovim/ui/screen.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    def clear(self):
+        """Clear the screen."""
+        self._clear_region(0, self.rows - 1, 0, self.columns - 1)
+
+    def eol_clear(self):
+        """Clear from the cursor position to the end of the scroll region."""
+        try:
+            self._clear_region(self.row, self.row, self.col, self.right)
+        except IndexError:
+            print self.row, self.col, self.right
```
Comment (fwalch) [inline neovim/ui/screen.py]: Braces are missing here (for Python 3).

Comment (tarruda) [inline neovim/ui/screen.py]: This is just a debug print, nvim should not send invalid indexes. Removed now

**Comment by Tranquility** [comment]
It still happens. I tried to locate the problem for hours. I even tried to break your vimrc (which worked fine). Please try mine and see whether you can replicate the problem. https://gist.github.com/Tranquility/6074b37b33196a2ee26c

**Comment by tarruda** [comment]
@Tranquility I've fixed the tickit UI to render default foreground/background properly, but you need to update this and neovim/neovim#1657

**Comment by Tranquility** [comment]
The cmd-line bg color is still wrong and also if I open a file directly the part where no text is has a wrong background. 
neovim-issue

If I call `:redraw!` after opening a file everything works.  But if I open pynvim without a file, open a file afterwards and call then `:redraw!` I get the previously described ui glitch.
neovim-ui

**Comment by tarruda** [comment]
> The cmd-line bg color is still wrong and also if I open a file directly the part where no text is has a wrong background. 

If you are using [this colorscheme](https://github.com/w0ng/vim-hybrid/blob/master/colors/hybrid.vim), it won't work without modification because tests like `if t_Co` or `if has('gui_running')` don't make sense in this new UI architecture(it is possible to have multiple UIs attached, and they can be TUIs or GUIs)

> But if I open pynvim without a file, open a file afterwards and call then :redraw! I get the previously described ui glitch.

I'm getting a lot of problems with your vimrc and color scheme, will investigate tomorrow

**Comment by fwalch** [comment]
> the part where no text is has a wrong background

I get this with `haya14busa/incsearch.vim`. To reproduce, setting up the plugin, adding `map /  <Plug>(incsearch-forward)` to the `.nvimrc` and typing `/` should be enough.

62724644-5f1e-400a-b66f-6de942bb718c

**Comment by tarruda** [comment]
@Tranquility do you get redraw glitches if ultisnips is disabled?

**Comment by Tranquility** [comment]
@tarruda only disabling ultisnips was not enough but after disabling ultisnips AND youcompleteme the problem seems to be gone.

**Comment by Tranquility** [comment]
@tarruda everything works now!

**Comment by tarruda** [comment]
> @tarruda everything works now!

Even the background issue?

**Comment by Tranquility** [comment]
Ok, I found something else. Maybe it is to early for that but the shell command output is very slow and I can't terminate it with ctrl-c.

I found that out by using fugitives :Ggrep with a pattern that has a lot of matches.

**Comment by Tranquility** [comment]
> Even the background issue?

Yes, that is fixed. Only the command line bg color is wrong before the first redraw.

**Comment by Tranquility** [comment]
Ok, looks like it's already tracked by https://github.com/neovim/neovim/issues/1234? But it still takes a really long time to display the output.

**Comment by tarruda** [comment]
@Tranquility did the last commit improve command output speed for you?

**Comment by Tranquility** [comment]
@tarruda no it is still very slow. I just execute :Glog in the neovim folder and it takes more than 10 seconds.

**Comment by tarruda** [comment]
> @tarruda no it is still very slow. I just execute :Glog in the neovim folder and it takes more than 10 seconds.

Unfortunately there's not much I can do right now, the bottleneck is in the internal scroll function which is implemented in pure python. Here are the initial entries of the profiler output:

```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     2679   13.720    0.005   26.644    0.010 screen.py:68(scroll)
 19237899   12.514    0.000   12.514    0.000 screen.py:27(copy)
       37    4.524    0.122    4.524    0.122 {method 'poll' of 'select.epoll' objects}
      233    2.231    0.010    2.231    0.010 {_cffi__xeb0d4df1xe5c9a0b5.tickit_term_flush}
    86749    1.388    0.000    1.646    0.000 screen.py:105(iter)
 ```

It should run much faster once its ported to C

**Pull request merged**